File: 11-container-with-most-water/11-container-with-most-water.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        # a container needs at least two walls, aka length of height = 2
        
        l = 0
        r = len(height)-1
        maxArea = -(sys.maxint)
        while l < r:
            currentArea = min(height[l],height[r]) * (r - l)
            maxArea = max(currentArea, maxArea)
            
            if height[l] > height[r]:
                r-=1
            elif height[l] < height[r]:
                l+=1
            else:
                r-=1
                l+=1
        
        return maxArea

File: 11-container-with-most-water/11-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    // length = diff between 2 indices
    // area = length x height
    // height = min(height[i], heigh[j])
    
    // start at the two ends
    // two pointers: beg,end
    
    // whichever side is smaller, move pointer
    // if same, move both
    
    let maxArea = 0
    let l = 0
    let r = height.length - 1
    
    // O(n)
    while (l < r) {
        let currentArea = Math.min(height[l],height[r]) * (r - l)
        maxArea = Math.max(maxArea, currentArea)
        
        if (height[l] < height[r]) {
            l ++
        } else if (height[l] > height[r]) {
            r --
        } else {
            l ++
            r --
        }
    }
    
    return maxArea
};